xls/synthesis/timing_characterization_client_main.py

```diff
@@ -54,85 +88,304 @@ def _synth(stub: synthesis_service_pb2_grpc.SynthesisServiceStub,
+  add_op_model.estimator.regression.factors.add(
+    source=delay_model_pb2.DelayFactor.OPERAND_BIT_COUNT,
+    operand_number=0)
+
+  op = ENUM2NAME_MAP[kop]
+
+  # Compute samples
+  results = []
+  for bit_count in [1] + list(range(2, MAX_WIDTH, 2)):
+    op_type = f'bits[{bit_count}]'
+    ir_text = op_module_generator.generate_ir_package(
+      op, op_type, (op_type, op_type))
 
+    module_name = f'{op}_{bit_count}'
+    mod_generator_result = op_module_generator.generate_verilog_module(module_name, ir_text)
+    top_name = module_name + '_wrapper'
+    verilog_text = op_module_generator.generate_parallel_module([mod_generator_result], top_name)
+    try:
+      result = _synth(stub, verilog_text, top_name)
+    except grpc.RpcError:
+      break
 
-def _run_binop_and_add(
-    op: str, kop: str, model: delay_model_pb2.DelayModel,
+    ps = 1e12/result.max_frequency_hz
+    result = delay_model_pb2.DataPoint()
+    result.operation.op = kop
+    result.operation.bit_count = bit_count
+
+    operand = result.operation.operands.add()
+    operand.bit_count = bit_count
+
+    result.delay = int(ps)
+    results.append(result)
+  model.data_points.extend(results)
+
+
+def _run_extension(
+    kop: str, model: delay_model_pb2.DelayModel,
     stub: synthesis_service_pb2_grpc.SynthesisServiceStub) -> None:
-  """Runs characterization for the given binop and adds it to the model."""
+  # Add op_model to protobuf message
   add_op_model = model.op_models.add(op=kop)
   add_op_model.estimator.regression.factors.add(
-      source=delay_model_pb2.DelayFactor.OPERAND_BIT_COUNT, operand_number=0)
-  model.data_points.extend(_run_binop(op, kop, stub))
+    source=delay_model_pb2.DelayFactor.OPERAND_BIT_COUNT,
+    operand_number=0)
 
+  op = ENUM2NAME_MAP[kop]
 
-def run_characterization(
+  # Compute samples
+  widths = [1] + list(range(2, MAX_WIDTH, 2))
+  new_bit_counts = list(range(2, MAX_WIDTH, 16))
+  combs = product(widths, new_bit_counts)
+  results = []
+  for bit_count, new_bit_count in combs:
+    op_type = f'bits[{bit_count}]'
+    ret_type = f'bits[{new_bit_count}]'
+    ir_text = op_module_generator.generate_ir_package(
+      op, ret_type, (op_type,),
+      attributes=[('new_bit_count', new_bit_count)])
+
+    module_name = f'{op}_{bit_count}_{new_bit_count}'
+    mod_generator_result = op_module_generator.generate_verilog_module(module_name, ir_text)
+    top_name = module_name + '_wrapper'
+    verilog_text = op_module_generator.generate_parallel_module([mod_generator_result], top_name)
+    try:
+      result = _synth(stub, verilog_text, top_name)
+    except grpc.RpcError:
+      break
+
+    ps = 1e12/result.max_frequency_hz
+    result = delay_model_pb2.DataPoint()
+    result.operation.op = kop
+    result.operation.bit_count = new_bit_count
+
+    operand = result.operation.operands.add()
+    operand.bit_count = bit_count
+
+    operand = result.operation.operands.add()
+    operand.bit_count = 64
```

Comment:
The extension operands (sext and zext) only have a single operand. They have a constant value attribute new_bit_count, but that is not an operand.

Reply:
I see, how should I handle that? Is it a literal operand? Should I use HAS_LITERAL_OPERAND?
I'm not sure if these extension operations actually result in a delay.